lenskit-api/src/main/java/org/lenskit/api/ItemRecommender.java

```diff
@@ -0,0 +1,133 @@
+ * specifying the exclude set, however, so applications with particular needs in
+ * this respect can manually provide the sets they need respected.
+ *
+ * ## Ordering
+ *
+ * If the recommender has an opinion about the order in which recommendations should be displayed,
+ * the result set will present items in that order.  For many recommenders, this will be descending order
+ * by score; however, this interface does not guarantee a relationship between scores and ordering.
+ *
+ * @compat Public
+ */
+public interface ItemRecommender {
+    /**
+     * Recommend all possible items for a user using the default exclude set.
+     *
+     * @param user The user ID.
+     * @return The recommended items.
+     * @see #recommend(long, int, Set, Set)
+     */
+    ResultList<Result> recommend(long user);
+
+    /**
+     * Recommend up to <var>n</var> items for a user using the default exclude
+     * set.
+     *
+     * @param user The user ID.
+     * @param n    The number of recommendations to return.
+     * @return The recommended items.
+     * @see #recommend(long, int, Set, Set)
+     */
+    ResultList<Result> recommend(long user, int n);
+
+    /**
+     * Recommend all possible items for a user from a set of candidates using
+     * the default exclude set.
+     *
+     * @param user       The user ID.
+     * @param candidates The candidate set (can be null to represent the
+     *                   universe).
+     * @return The recommended items.
+     * @see #recommend(long, int, Set, Set)
+     */
+    ResultList<Result> recommend(long user, @Nullable Set<Long> candidates);
+
+    /**
+     * Produce a set of recommendations for the user. This is the most general
+     * recommendation method, allowing the recommendations to be constrained by
+     * both a candidate set and an exclude set. The exclude set is applied to
+     * the candidate set, so the final effective candidate set is
+     * <var>canditates</var> minus <var>exclude</var>.
+     *
+     * @param user       The user's ID
+     * @param n          The number of ratings to return. If negative, there is
+     *                   no specific recommendation list size requested.
+     * @param candidates A set of candidate items which can be recommended. If
+     *                   {@code null}, all items are considered candidates.
+     * @param exclude    A set of items to be excluded. If {@code null}, a default
+     *                   exclude set is used.
+     * @return A list of recommended items. If the recommender cannot assign
+     *         meaningful scores, the scores will be {@link Double#NaN}. For
+     *         most scoring recommenders, the items will be ordered in
+     *         decreasing order of score. This is not a hard requirement — e.g.
+     *         set recommenders are allowed to be more flexible.
+     */
+    ResultList<Result> recommend(long user, int n, @Nullable Set<Long> candidates,
```

Comment:
Theoretically these "without details" versions could simply return List<Long> for people who just want plain old java.